rfcs/taskforce/0003-Integration-test-infrastructure-spec.md

```diff
@@ -0,0 +1,214 @@
+
+A network manager is basically a wrapper around the network infrastructure using which a testnet is deployed. The network manager implemented in ocaml would take as input a network configuration and use it to generate required configuration for the deployment and deploy a testnet. The network manager is responsible for creating the seed nodes that other nodes can connect to. A separate network manager would be created for each environment where we want to run the testnet. The choice of a network manager is made by the test executive based on a user input (whether to run locally or on the cloud). The interface between a network manager and the test executive exposes all the nodes and the commands that can be issued to the nodes.
+
+In the current testnet infrastructure, a network manager would generate the necessary terraform configuration files and deploy using terraform.
+To run it locally, using minikube perhaps, we would create another network manager that deploys a testnet locally.
+
+The network manager itself can be specified as:
+
+```ocaml
+
+module Node : Node_intf
+(* Deploy the network*)
+val deploy : Network_config.t -> Daemon_config.t -> Testnet.t Deferred.t
+
+(*Tear down the network*)
+val destroy : Testnet.t -> unit Deferred.t
+```
+
+```ocaml
+
+module type Node_intf : sig
+    type t
+
+    val node_id_json : t -> json
+
+    val start : t -> unit Deferred.t
+
+    val stop : t -> unit Deferred.t
+
+    ...
+end
+
+module type Testnet = sig
+
+    type t =
+        {
+          block_producers : Node.t list
+        ; snark_coordinators: Node.t list
+        ; archive_nodes: Node.t list
+        ; ... }
+end
+```
+
+A node will encapsulate all the information needed to connect to it and make requests such as sending payments, retrieving a node's status, and so on. These requests implemented by the network manager could be using graphql commands. Given our intention to make all the cli commands use graphql queries, cli commands that do not currently have a graphql query should have that implemented first.
+Also, care should be taken to not add any test-specific queries/mutations/subscriptions that would not be used by users.
+
+The interface should also implement functions to start and stop a node that can be triggered at any time by the test program. This is particularly needed for testing functionalities like bootstrap and catchup. The network infrastructure should facilitate stopping and restarting coda daemon containers (with or w/o persisting the config directory). Currently a pod or a container within a pod is restarted automatically every time the daemon terminates. This behavior should be changed to support restarting a node on request. The following are some of the options to enable restart on request:
+    1. Have another process as the root of the container that initiates or terminates coda daemon based on a request from the network manager. This requires implementing and maintaining another application.
```

Comment:
+1 - #1 has the drawback of maintaining another application (or group of apps) for sure though it lends itself pretty nicely  and better imo (than the other 2) to both manual/on-demand and automated restarts or even other types of node manipulation.

Reply:
One other option that might be worth considering is a *kubernetes* [job](https://kubernetes.io/docs/concepts/workloads/controllers/job/). Somewhat torn between this and #1 though I kind of like the idea of jobs due to their:
- separation/independence from the testnet infrastructure deployment process, itself (jobs can be deployed at any time against any testnet without dependence on a supporting *puppet-master* service
- generally consist of clearly defined end or goal states
- removes single point of failure (e.g. bug introduced by newly added *puppeteer'ing* operation would be confined within job rather than *puppet-master* service)
- can be applied using existing infra
- built-in scaling, parallelization and scheduling features
- potential for dynamic job creation (e.g. simulating conditional, branching test scenarios)

Reply:
Based on our discussion, having kubernetes jobs to terminate the daemon container (with `exit 0`) would require redeploying the testnet (or configure singleton pods?) to start the pod again. Also, we'd want restart policy to be set to `never` because container or pods dying should result in a test failure because the intended behavior would not have been tested in which case the test should be rerun.
If we redeploy then we'll need `StatefulSets` for persistent storage which is currently not implemented (so bump the priority?).
I like option 1 slightly more because we have both in one container with minimal external dependencies and therefore can work quite well if we were to connect multiple containers on the same host (for running the test locally)

Reply:
> Based on our discussion, having kubernetes jobs to terminate the daemon container (with `exit 0`) would require redeploying the testnet (or configure singleton pods?) to start the pod again

Any option we use (* dedicated/per daemon puppet-master pod, * single/stand-alone puppet-master pod or * a job pod), will require redeploying or re-installing the *testnet* in order to re-start the *daemon* pod if the *daemon* container exits with any code (0 or otherwise) afaia. In other words, there is no concept of restarting containers in pods within k8s outside of a *RestartPolicy* specification defined at deploy time.

e.g. something like: `kubectl stop/restart <pod-name> --container <container-name>` does NOT exist.

There are, however, "hacks" around this limitation by for example, `scaling` a *pod* to 0 and then re-scaling to 1; in a sense simulating stopping and starting ALL pod containers. The key point here is that ALL containers are restarted which may or may not be a desirable effect.

Reply:
IMO if we're trying to avoid *deleting/re-installing pods* due to the `StatefulSet` requirement, something along the lines of `exec`ing into individual *coda-daemon* containers and interacting directly with the coda daemon process in a way that prevents its container from ever stopping is really what we're working with.

Haven't tested it out yet but we may be able to just send *suspend/continue* signals to the coda daemon process to simulate stops/restarts a la something like:
- suspend/stop: `kubectl exec -i -t fish-producer-1 --container coda-daemon -- kill -TSTP <daemon-pid>`
- continue/restart: `kubectl exec -i -t fish-producer-1 --container coda-daemon -- kill -CONT <daemon-pid>`

Using a technique like this though, any of the previously mentioned options could be used with pretty much the biggest advantage between the 3 coming in the form of saved or more efficient use of cluster system resources (i.e. option # 1 (dedicated puppet-master) being the most wasteful and # 3 (dedicated puppet job) being the least wasteful for obvious reasons and as discussed earlier.

Reply:
There also seem to be the following technical concerns:
- asynchronicity
- multi-processing/threading
- scalability 

...when deciding on the implementation of the *puppet-master*. In regards to all three, I'd have to say that jobs again provide major benefits compared to the other 2 (or really 1 -- puppet-master service) in the sense that a "puppeteering" script could be written in any language (OCaml, python, Go, etc) w/o regards to the above concerns as *k8s*' jobs construct would manage them.

Reply:
> I like option 1 slightly more because we have both in one container with minimal external dependencies and therefore can work quite well if we were to connect multiple containers on the same host (for running the test locally)

Mostly agreed though there should be the same amount of external dependencies (either a puppet-master service or puppet-master script) between the options with a script being simpler to maintain IMO.

Also to note, jobs are supported by `minikube` afaik and can be triggered on a local k8s cluster or triggered from a developer's laptop or whatever to act on a remote *k8s* cluster (e.g. one created in *GKE*) using something like:

```
kubectl apply -f https://github.com/CodaProtocol/coda/test/jobs/restart_pod.yml
```